### PR TITLE
Addition of softmin activation function

### PR DESCRIPTION
**Pull request by Aakash-kaushik**
## Addition of softmin activation function in ann/layer, 
- Adding the Softmin activation function in ann/layers

### CODE REVIEW

**Comment by Aakash-kaushik** [comment]
The derivative of Softmin function that i did can be accessed [here](https://drive.google.com/file/d/12hdnj7mv4rVXf__oGMh5VGpYreus3kSO/view) .

**Comment by kartikdutt18** [comment]
I think the derivative is correct. Is the output not matching?

**Comment by kartikdutt18** [comment]
Also for really small values we can use catch's version of BOOST_SMALL instead.

**Comment by kartikdutt18** [comment]
Also, This does have a the derivative of softmin. Here is the [link](https://core.ac.uk/download/pdf/211842025.pdf). Take a look at the first search of softmin or Page 9.

File: src/mlpack/methods/ann/layer/softmin_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename OutputType>
+void Softmin<InputDataType, OutputDataType>::Forward(
+    const InputType& input,
+    OutputType& output)
+{ 
+  InputType inputMin = arma::repmat(arma::min(input,1),1,input.n_cols);
+  output = arma::repmat(arma::log(arma::sum(
+      arma::exp(- (input - inputMin)),1)), 1, input.n_cols);
+  output = arma::exp(- (input - inputMin) - output);
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/softmin_impl.hpp:36]:
```suggestion
  output = arma::exp(-(input - inputMin) - output);
```

File: src/mlpack/methods/ann/layer/softmin.hpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+/**
+* Implementation of the Softmin layer. The Softmin function takes as a input
+* a vector of K real numbers, rescaling them so that the elements of the
+* K-dimensional output vector lie in the range [0, 1] and sum to 1.
+* 
+* @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+*         arma::sp_mat or arma::cube).
+* @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+*         arma::sp_mat or arma::cube).
+*/
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/softmin.hpp:30]:
```suggestion
*  Implementation of the Softmin layer. The Softmin function takes as a input
*  a vector of K real numbers, rescaling them so that the elements of the
*  K-dimensional output vector lie in the range [0, 1] and sum to 1.
* 
*  @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
*          arma::sp_mat or arma::cube).
*  @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
*          arma::sp_mat or arma::cube).
*/
```

Should align with the first line.

**Comment by Aakash-kaushik** [comment]
> Also for really small values we can use catch's version of BOOST_SMALL instead.

Hey, i think in floating point comparison section on [this](https://github.com/catchorg/Catch2/blob/master/docs/assertions.md#top) page form catch2,  Scale is the closest to BOOST_SMALL, I am not sure if catch2 offers a direct small comparison, either i could decrease the tolerance with epsilon that is currently set to 1e-5 or either use scale with a small value.
What should i go with ?

**Comment by Aakash-kaushik** [comment]
> I think the derivative is correct. Is the output not matching?

Even if the derivative is right i am not able to make it in a form similar to [this](https://github.com/mlpack/mlpack/blob/97a5cee930bf80251c8c9359b8500adccaafad17/src/mlpack/methods/ann/layer/softmax_impl.hpp#L46) and i am trying to keep it similar because i don't think the derivative of softmin will change much from softmax instead of a minus sign.

**Comment by Aakash-kaushik** [comment]
Hey @zoq, 
I have implemented the derivative but i am really unsure, can you please take a look.

File: src/mlpack/tests/activation_functions_test.cpp [hunk 3]
```diff
@@ -1063,3 +1113,29 @@ TEST_CASE("GaussianFunctionTest", "[ActivationFunctionsTest]")
...
+  const arma::colvec desiredDerivatives("1 1 1 1 1 1 1 0");
+  */
+
+  CheckSoftminActivationCorrect(activationData,
+                                desiredActivations);
+  /* 
+  Commented out because derivatives haven't been figured out.
+  CheckHardShrinkDerivativeCorrect(desiredActivations,
+                                   desiredDerivatives);
+  */
```
Comment (kartikdutt18) [inline src/mlpack/tests/activation_functions_test.cpp:1140]: Could you add tests for the backward as well.

Comment (Aakash-kaushik) [reply src/mlpack/tests/activation_functions_test.cpp]: Done.

Comment (Aakash-kaushik) [reply src/mlpack/tests/activation_functions_test.cpp]: > Could you add tests for the backward as well.

    Can you take a look now, I am still unsure about the Backward and it's tests.

Comment (kartikdutt18) [reply src/mlpack/tests/activation_functions_test.cpp]: Sure, I'll take a look tonight.

**Comment by Aakash-kaushik** [comment]
The Memory leak failure seems to be in the Qlearning test, and not related to this.

**Comment by Aakash-kaushik** [comment]
Hey @birm, Can you review this?

File: src/mlpack/tests/activation_functions_test.cpp [hunk 4]
```diff
@@ -1063,3 +1113,21 @@ TEST_CASE("GaussianFunctionTest", "[ActivationFunctionsTest]")
...
+
+/**
+ * Basic test of the Softmin function.
+ */
+TEST_CASE("SoftminFunctionTest", "[ActivationFunctionsTest]")
+{
+  const arma::colvec activationData1("1.7 3.6");
+
+  // Hand-calculated values.
+  const arma::colvec desiredActivations("0.86989153 0.13010847");
```
Comment (birm) [inline src/mlpack/tests/activation_functions_test.cpp:1125]: :+1: matches pytorch reference
```python
>>> a = torch.tensor([1.7,3.6])
>>> m(a)
__main__:1: UserWarning: Implicit dimension choice for softmin has been deprecated. Change the call to include dim=X as an argument.
tensor([0.8699, 0.1301])
>>> o = m(a)
>>> o.grad_fn(torch.tensor([1.0,0.0]))
tensor([ 0.1132, -0.1132], grad_fn=<SoftmaxBackwardDataBackward>)

```

**Review by birm — APPROVED** [review]
Looks good as far as I can tell :)

File: src/mlpack/tests/activation_functions_test.cpp (hunk 4 above)
Comment (Aakash-kaushik) [reply src/mlpack/tests/activation_functions_test.cpp:1125]: I am still a bit suspicious about the backward function so i will check with some more values and post the results here, also thanks for reviewing this so quickly :)

Comment (Aakash-kaushik) [reply src/mlpack/tests/activation_functions_test.cpp:1125]: Hey @birm,
with the input of `4.2, 2.4, 7.0, 6.4` and taking the derivative WRT `1.0, 0.0, 1.0, 0.0`
PyTorch produces : `0.1181, -0.1231,  0.0072, -0.0023`
```
>>> a = torch.tensor([4.2, 2.4, 7.0, 6.4],requires_grad=True)
>>> o = m(a)
>>> o
tensor([0.1385, 0.8378, 0.0084, 0.0153], grad_fn=<SoftmaxBackward>)
>>> o.grad_fn(torch.tensor([1.0,0.0,1.0,0.0]))
tensor([ 0.1181, -0.1231,  0.0072, -0.0023],
       grad_fn=<SoftmaxBackwardDataBackward>)
```
Whereas this implementation produces : `-0.1181, 0.1231, -0.0072, 0.0023`
So changing the current backward from : `g = input % (arma::repmat(arma::sum(gy % input), input.n_rows, 1) - gy);`
To this : `g = input % (gy - arma::repmat(arma::sum(gy % input), input.n_rows, 1));`
inverts the minus sign and so we can get the correct output as per PyTorch.
So i will go ahead and update the backward function here.

Comment (Aakash-kaushik) [reply src/mlpack/tests/activation_functions_test.cpp:1125]: Also i am surprised that this and the `Softmax` backward are the exact same.

**Review by birm — CHANGES_REQUESTED** [review]
Waiting until you finish the catch migration and builds resolve

**Comment by Aakash-kaushik** [comment]
> Waiting until you finish the catch migration and builds resolve

The catch migration wasn't for this branch, But I probably forgot to change the branch yesterday night and pushed the changes today. The builds will resolve now and I will create a different PR for the catch migration.

**Comment by Aakash-kaushik** [comment]
The test case was wrong now it is corrected according to PyTorch. 
```python
tensor([1.7000, 3.6000], requires_grad=True)
>>> o = m(a)
<stdin>:1: UserWarning: Implicit dimension choice for softmin has been deprecated. Change the call to include dim=X as an argument.
>>> o
tensor([0.8699, 0.1301], grad_fn=<SoftmaxBackward>)
>>> o.grad_fn(torch.tensor([1.0,0.0]))
tensor([ 0.1132, -0.1132], grad_fn=<SoftmaxBackwardDataBackward>)
```

**Comment by Aakash-kaushik** [comment]
Hey @birm, the failing checks doesn't seem to be because of the changes in this PR but i am not sure can you check ?

File: src/mlpack/tests/activation_functions_test.cpp (hunk 4 above)
Comment (birm) [reply src/mlpack/tests/activation_functions_test.cpp:1125]: Please add the new test case here

Comment (Aakash-kaushik) [reply src/mlpack/tests/activation_functions_test.cpp]: Done.

**Comment by Aakash-kaushik** [comment]
The failure of tests again doesn't looks related to this PR, also i was curious about the mac os python test failure does that occasionally happen ?

**Comment by Aakash-kaushik** [comment]
@birm, @zoq I think this is ready for a final review.

**Comment by kartikdutt18** [comment]
Hey @Aakash-kaushik, following up on your comment, I have already fixed the threshold issue.
Also, I have re-run the tests.

**Comment by kartikdutt18** [comment]
Hey @Aakash-kaushik, Could you rebase the branch. We should get a green build.

**Comment by Aakash-kaushik** [comment]
> Hey @Aakash-kaushik, Could you please rebase the branch. We should get a green build.

Hi @kartikdutt18, It's done.

**Comment by Aakash-kaushik** [comment]
Windows VS16 Plain failure fixed in #2639 .

**Review by birm — APPROVED** [review]
Awesome; great work, and thanks for checking past my initial validation attempt :slightly_smiling_face:

**Review by kartikdutt18 — APPROVED** [review]
Looks good to me as well. Bringing it in.

**Comment by Aakash-kaushik** [comment]
> Awesome; great work, and thanks for checking past my initial validation attempt slightly_smiling_face

Doing it right. :smile:

File: src/mlpack/methods/ann/layer/softmin_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename OutputType>
+void Softmin<InputDataType, OutputDataType>::Forward(
+    const InputType& input,
+    OutputType& output)
+{ 
+  InputType inputMin = arma::repmat(arma::min(input,0), input.n_rows, 1);
+  output = arma::repmat(arma::log(arma::sum(
+      arma::exp(-(input - inputMin)),0)), input.n_rows, 1);
+  output = arma::exp(-(input - inputMin) - output);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/softmin_impl.hpp:36]: Do you have a reference for this? I'm not sure where the `log` comes in. If I remember right the definition of Softmin is:

`Softmin(x) = exp(-x) / sum(exp(-x))`

Maybe I missed something?

Comment (zoq) [reply src/mlpack/methods/ann/layer/softmin_impl.hpp:36]: When I rescale the output to be in the range (0, 1), I would do:

`Softmin(x) = exp(-x - max(x)) / sum(exp(-x - max(x)))`

But I still don't see the need for `log`.

Comment (Aakash-kaushik) [reply src/mlpack/methods/ann/layer/softmin_impl.hpp:36]: Hey @zoq you are right with the definition of Softmin and generally we scale Softmin with `min(x)`(the scaling is with min(x) in PyTorch too) so it would be something like `Softmin(x) = exp(-x - min(x)) / sum(exp(-x - min(x)))` And for why and where the log came from so i looked at the definition of [Softmax](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/softmax_impl.hpp#L34) in mlpack, That defination is derived by taking log on both sides of Softmax so i implemented it in a similar way

1. `log(Softmin(x)) = log(exp(-x - min(x))) - log(sum(exp(-x - min(x))))`      // Taking the log on both sides.
2. `log(Softmin(x)) = (-x - min(x)) - log(sum(exp(-x - min(x))))`                   // The log gets cancelled with the exponent in the term `log(exp(-x - min(x)))`.
3. `Softmin(x) = exp((-x - min(x)) - log(sum(exp(-x - min(x)))))`                  // The only log that remains is in the second term because of the summing operation.

And I think we implement it this way to reduce the computational cost. I haven't really done any benchmarks so i am not sure.

Let me know hat do you think about this.

Comment (zoq) [reply src/mlpack/methods/ann/layer/softmin_impl.hpp:36]: Right, same problem with the `Softmax` layer, I will go over the `Softmax`, but I don't think the implementation is correct here. We should get the same output as e.g. PyTorch:

```python
import torch.nn.functional as F
import torch as th

data = th.randn(4)

data[0] = 1.43508
data[1] = 0.8377550303
data[2] = 0.008420976
data[3] = 0.0153440186
print(data)
print(F.softmin(data, dim=0))

tensor([1.4351, 0.8378, 0.0084, 0.0153])
tensor([0.0899, 0.1635, 0.3746, 0.3720])
```

which is currently not the case. So unless you see a reason can you open another PR and fix the implementation?

Comment (Aakash-kaushik) [reply src/mlpack/methods/ann/layer/softmin_impl.hpp:36]: We are definitely getting the same forward and backward outputs as PyTorch: 
Our implementation:-
```c++
#include <iostream>
#include <armadillo>

using namespace std;

int main()
{
  arma::colvec input("1.43508 0.8377550303 0.008420976 0.0153440186");
  
  arma::colvec inputMin = arma::repmat(arma::min(input,0),input.n_rows,1);
  arma::colvec output = arma::repmat(arma::log(arma::sum(
      arma::exp(-(input - inputMin)),0)),input.n_rows,1);
  output = arma::exp(-(input - inputMin) - output);
  output.print("output");


  arma::colvec gy = arma::ones<arma::mat>(output.n_elem);
  gy(1) = 0;
  gy(3) = 0;
  arma::colvec g = output % (gy - (arma::repmat(arma::sum(gy % output,0), output.n_rows, 1)));
  g.raw_print("g");
  return 0;
}
```
output
  ` 0.0899
   0.1635
   0.3746
   0.3720`
g
`0.0481615
-0.0759292
0.200581
-0.172813`

PyTorch:-
```python
import torch
import torch.nn as nn

data = torch.tensor([1.43508, 0.8377550303, 0.008420976, 0.0153440186], requires_grad = True) 

softmin = nn.Softmin()
out = softmin.forward(data)
print("forward" + str(out))
print("derivative: " + str(out.grad_fn(torch.tensor([1.0, 0.0, 1.0, 0.0]))))
```
`forwardtensor([0.0899, 0.1635, 0.3746, 0.3720], grad_fn=<SoftmaxBackward>)`
`derivative: tensor([ 0.0482, -0.0759,  0.2006, -0.1728], grad_fn=<SoftmaxBackwardDataBackward>)`

Let me know if i am still overlooking something.

Comment (Aakash-kaushik) [reply src/mlpack/methods/ann/layer/softmin_impl.hpp:36]: Just a addition: 
```python
torch.nn.functional.softmin()
```
```python
torch.nn.Softmin()
```
They both produce the same output, I think the first a function or a object and the later is a class.

Comment (zoq) [reply src/mlpack/methods/ann/layer/softmin_impl.hpp:36]: Ahh, I run my test with the expected output as input, okay, great the output is the same. So the implementation is correct, great.

Comment (zoq) [reply src/mlpack/methods/ann/layer/softmin_impl.hpp:36]: But I'm still wondering if the `log` approach is the best here especially in combination with `repmat`. `max` and `each_col` should be faster compared with creating a temporary matrix.

Comment (Aakash-kaushik) [reply src/mlpack/methods/ann/layer/softmin_impl.hpp:36]: We can do one thing. I will setup a testing environment with both the implementations and then I can generate some huge column vectors and run the tests multiple time for both the implementations to get a average reading. can you guide me to some tools for this. i haven't really bench-marked anything in C++

Comment (zoq) [reply src/mlpack/methods/ann/layer/softmin_impl.hpp:36]: You could use `chrono` - https://en.cppreference.com/w/cpp/chrono for the timing.

Comment (Aakash-kaushik) [reply src/mlpack/methods/ann/layer/softmin_impl.hpp:36]: Hey @zoq, I am back with the results you can see the testing code for the current implementation [here](https://gist.github.com/Aakash-kaushik/a51a1543ef76b8c0a8285e5ce51c5de4) and for the one you suggested [here](https://gist.github.com/Aakash-kaushik/d877015f971001787fd2e43f0b2596c9) and yes you were right the implementation you suggested is considerably fast. 
execution time for log implementation: `222150`
execution time for implementation you suggested: `149912`
Approx `48%` faster than the current implementation.

So I will go ahead and open a PR for both Softmin and Softmax. If that's fine ?

Comment (zoq) [reply src/mlpack/methods/ann/layer/softmin_impl.hpp:36]: Sounds great, thanks for looking into the different implementations.

Comment (Aakash-kaushik) [reply src/mlpack/methods/ann/layer/softmin_impl.hpp:36]: Hey, I have created a PR: #2657 regarding this, give it a look when you are free.
